addOns/authhelper/src/main/java/org/zaproxy/addon/authhelper/internal/ClientSideHandler.java

```diff
@@ -64,81 +78,208 @@ public void handleMessage(HttpMessageHandlerContext ctx, HttpMessage msg) {
-                // to find one of them
-                for (SessionToken st : reqSessionTokens) {
-                    smReqDetails = AuthUtils.findSessionTokenSource(st.getValue(), firstHrefId);
-                    if (smReqDetails != null) {
-                        authMsg = smReqDetails.getMsg();
-                        LOGGER.debug("Session token found in href {}", getHrefId(authMsg));
-                        break;
-                    }
+        for (SessionToken st : unkSessionTokens) {
+            // See if we can find the reqs for the unknown session tokens, then see if they are
+            // better than the current one
+            SessionManagementRequestDetails smReqDetails =
+                    AuthUtils.findSessionTokenSource(st.getValue(), firstHrefId);
+            if (smReqDetails != null) {
+                candidate = new AuthRequestDetails(msg);
+                if (candidate.isBetterThan(authReq, headerConfigs)) {
+                    LOGGER.debug(
+                            "Found better auth candidate {} {}",
+                            msg.getRequestHeader().getMethod(),
+                            msg.getRequestHeader().getURI());
+                    authReq = candidate;
                 }
             }
-
-            if (authMsg != null && View.isInitialised()) {
-                AuthUtils.logUserMessage(
-                        Level.INFO,
-                        Constant.messages.getString(
-                                "authhelper.auth.method.browser.output.sessionid", getHrefId(msg)));
-            }
-        }
-        if (firstHrefId == 0 && msg.getHistoryRef() != null) {
-            firstHrefId = msg.getHistoryRef().getHistoryId();
         }
     }
 
-    private String getHrefId(HttpMessage msg) {
-        if (msg.getHistoryRef() != null) {
-            return Integer.toString(msg.getHistoryRef().getHistoryId());
-        }
-        return "?";
-    }
-
     public HttpMessage getAuthMsg() {
-        return authMsg;
+        return authReq.getMsg();
     }
 
     public void resetAuthMsg() {
-        this.authMsg = null;
+        this.authReq = null;
+    }
+
+    protected static boolean isBetterThan(
+            SessionManagementRequestDetails smrd1, SessionManagementRequestDetails smrd2) {
+        if (smrd2 == null) {
+            return true;
+        }
+        if (smrd1.getConfidence() > smrd2.getConfidence()) {
+            return true;
+        }
+        if (smrd1.getConfidence() < smrd2.getConfidence()) {
+            return false;
+        }
+        return smrd1.getTokens().size() > smrd2.getTokens().size();
     }
 
     public HttpMessage getFallbackMsg() {
         return fallbackMsg;
     }
+
+    protected static List<Pair<String, String>> extractKeyValuePairs(String input) {
+        List<Pair<String, String>> keyValuePairs = new ArrayList<>();
+        Pattern pattern = Pattern.compile("\\{%([^:]+):([^%]+)%}");
```

Comment:
How often is this done? Would it be better to have this as a constant vs compiling on every use ?

Reply:
V good point!